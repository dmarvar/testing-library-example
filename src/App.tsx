import { useEffect, useState } from "react";
import axios from "axios";

import "./App.css";

const fakeApiCall = async () => {
  return axios.get<string>("/countToThree");
};

function App() {
  const [count, setCount] = useState(0);
  const [data, setData] = useState<string | undefined>(undefined);

  const onNumberThree = async () => {
    try {
      const result = await fakeApiCall();
      setData(result.data);
    } catch (e) {
      console.log(e);
    }
  };

  useEffect(() => {
    if (count === 3) {
      onNumberThree();
    }
  }, [count]);

  const addCount = () => {
    if (count < 3) {
      setCount(count + 1);
    }
  };

  return (
    <div className="App">
      <article className="App-article">
        {
          // No data escenario
          !data && (
            <>
              <h1>Counter</h1>
              <h2 data-testid="counter-test-id">{count}</h2>
              <br />
              <button onClick={addCount}>Increment count:</button>
            </>
          )
        }
        {
          // Data escenario
          data && <p>{data}</p>
        }
      </article>
    </div>
  );
}

export default App;
