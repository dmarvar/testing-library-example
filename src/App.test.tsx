import { logRoles, render, screen, waitFor } from "@testing-library/react";
import App from "./App";
import userEvent from "@testing-library/user-event";

describe("App", () => {
  it("should render correctly without changes", () => {
    const { baseElement, debug } = render(<App />);
    // eslint-disable-next-line testing-library/no-debugging-utils
    debug();

    logRoles(baseElement);

    expect(baseElement).toMatchSnapshot();

    expect.assertions(1);
  });

  it("should increment count until 3 and then show success message", async () => {
    render(<App />);

    const button = screen.getByRole("button");
    const count = screen.getByTestId("counter-test-id");

    expect(count.innerHTML).toBe("0");

    userEvent.click(button);
    expect(count.innerHTML).toBe("1");

    userEvent.click(button);
    expect(count.innerHTML).toBe("2");

    userEvent.click(button);
    expect(count.innerHTML).toBe("3");

    const finalText = await screen.findByText("Great Work!");

    await waitFor(() => expect(finalText).toBeInTheDocument());
  });
});
