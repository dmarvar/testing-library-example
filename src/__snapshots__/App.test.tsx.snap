// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly without changes 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <article
        class="App-article"
      >
        <h1>
          Counter
        </h1>
        <h2
          data-testid="counter-test-id"
        >
          0
        </h2>
        <br />
        <button>
          Increment count:
        </button>
      </article>
    </div>
  </div>
</body>
`;
